File: javascript/lib/pretestads.mjs
/**
 * Shared PreTestAds x402 client (Node 18+).
 *
 * Pays $5.00 USDC via x402 on Base to score one ad, then polls the free
 * status endpoint. Set EVM_PRIVATE_KEY (0x-hex; wallet needs USDC + a
 * little ETH on Base). The key signs locally and is never sent anywhere.
 *
 * (Paying on Solana from JS depends on your x402-fetch version's SVM
 * support — the Python scripts in ../python support Solana out of the box.)
 *
 * Requires: npm install x402-fetch viem
 */
import { readFileSync, existsSync } from "node:fs";
import { basename } from "node:path";
import { privateKeyToAccount } from "viem/accounts";
import { wrapFetchWithPayment } from "x402-fetch";

export const SCORE_URL =
  process.env.PRETESTADS_SCORE_URL ?? "https://pretestads.com/api/v1/score";

export function payingFetch() {
  const key = process.env.EVM_PRIVATE_KEY;
  if (!key) throw new Error("Set EVM_PRIVATE_KEY (Base wallet holding USDC).");
  const account = privateKeyToAccount(key);
  // Handles the 402 -> sign -> retry dance automatically.
  return wrapFetchWithPayment(fetch, account);
}

/** Pay $5 USDC and submit an ad (https URL or local file path). Returns the 202 body. */
export async function submitAd(media, windowSeconds = 15) {
  const fetchWithPay = payingFetch();

  let init;
  if (existsSync(media)) {
    const form = new FormData();
    form.append("file", new Blob([readFileSync(media)]), basename(media));
    form.append("window_seconds", String(windowSeconds));
    init = { method: "POST", body: form };
  } else {
    init = {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ media_url: media, window_seconds: windowSeconds }),
    };
  }

  const r = await fetchWithPay(SCORE_URL, init);
  if (r.status !== 202) {
    throw new Error(`Submit failed (${r.status}): ${(await r.text()).slice(0, 300)}`);
  }
  return r.json();
}

/** Poll the free status endpoint until complete/failed (~5-8 min typical). */
export async function waitForScore(statusUrl, { timeoutMinutes = 15, quiet = false } = {}) {
  const deadline = Date.now() + timeoutMinutes * 60_000;
  while (Date.now() < deadline) {
    const body = await (await fetch(statusUrl)).json();
    if (!quiet) console.log(`    ${new Date().toLocaleTimeString()}  status=${body.status}`);
    if (body.status === "complete" || body.status === "failed") return body;
    await new Promise((res) => setTimeout(res, 20_000));
  }
  throw new Error("Timed out waiting for the score.");
}

/** Pay, submit, wait. Returns the final result (score, label, verdict, timeseries). */
export async function scoreAd(media, windowSeconds = 15, opts = {}) {
  const job = await submitAd(media, windowSeconds);
  if (!opts.quiet) console.log(`Paid & queued: id=${job.id} (scoring takes ~5-8 min)`);
  return waitForScore(job.status_url, opts);
}

File: python/pretestads.py
"""Shared PreTestAds x402 client.

Pays $5.00 USDC via the x402 protocol (Solana or Base) to score one ad,
then polls the free status endpoint until the score is ready.

Wallet env vars (set at least one; keys sign locally, never sent anywhere):
    SOLANA_PRIVATE_KEY  base58 secret key (wallet needs USDC + a little SOL)
    EVM_PRIVATE_KEY     0x-hex key       (wallet needs USDC + ETH on Base)

Requires: pip install httpx "x402[evm,svm]"
"""
import asyncio
import os
import time

import httpx

SCORE_URL = os.environ.get("PRETESTADS_SCORE_URL", "https://pretestads.com/api/v1/score")


def _build_x402_client():
    from x402 import x402Client

    client = x402Client()
    registered = []

    if os.environ.get("SOLANA_PRIVATE_KEY"):
        from solders.keypair import Keypair
        from x402.mechanisms.svm.signers import KeypairSigner
        from x402.mechanisms.svm.exact import ExactSvmClientScheme

        signer = KeypairSigner(Keypair.from_base58_string(os.environ["SOLANA_PRIVATE_KEY"]))
        client.register("solana:*", ExactSvmClientScheme(signer))
        registered.append(f"solana ({signer.address})")

    if os.environ.get("EVM_PRIVATE_KEY"):
        from eth_account import Account
        from x402.mechanisms.evm.exact import ExactEvmClientScheme

        account = Account.from_key(os.environ["EVM_PRIVATE_KEY"])
        client.register("eip155:*", ExactEvmClientScheme(account))
        registered.append(f"evm ({account.address})")

    if not registered:
        raise RuntimeError("Set SOLANA_PRIVATE_KEY and/or EVM_PRIVATE_KEY to pay via x402.")
    return client, registered


def _media_kwargs(media: str, window_seconds: int):
    """Local file -> multipart upload; otherwise JSON media_url."""
    if os.path.isfile(media):
        with open(media, "rb") as f:
            return {
                "files": {"file": (os.path.basename(media), f.read(), "application/octet-stream")},
                "data": {"window_seconds": str(window_seconds)},
            }
    return {"json": {"media_url": media, "window_seconds": window_seconds}}


async def submit_ad(media: str, window_seconds: int = 15) -> dict:
    """Pay $5 USDC via x402 and submit an ad (URL or local path) for scoring.

    Returns the 202 body: {"id", "status", "status_url", "payment": {...}}.
    """
    from x402.schemas import PaymentRequired
    from x402.http.utils import encode_payment_signature_header

    client, wallets = _build_x402_client()

    async with httpx.AsyncClient(timeout=180) as http:
        # 1. Unpaid probe -> 402 with payment requirements
        r = await http.post(SCORE_URL, json={})
        if r.status_code != 402:
            raise RuntimeError(f"Expected 402, got {r.status_code}: {r.text[:300]}")
        payment_required = PaymentRequired.model_validate(r.json())

        # 2. Sign a USDC payment for whichever network matches our wallet
        payload = await client.create_payment_payload(payment_required)
        header = encode_payment_signature_header(payload)

        # 3. Paid retry with the actual media
        r = await http.post(
            SCORE_URL,
            headers={"PAYMENT-SIGNATURE": header},
            **_media_kwargs(media, window_seconds),
        )
        if r.status_code != 202:
            raise RuntimeError(f"Submit failed ({r.status_code}): {r.text[:300]}")
        return r.json()


async def wait_for_score(status_url: str, timeout_minutes: int = 15, quiet: bool = False) -> dict:
    """Poll the free status endpoint until complete/failed (~5-8 min typical)."""
    async with httpx.AsyncClient(timeout=60) as http:
        deadline = time.time() + timeout_minutes * 60
        while time.time() < deadline:
            body = (await http.get(status_url)).json()
            status = body.get("status")
            if not quiet:
                print(f"    {time.strftime('%H:%M:%S')}  status={status}", flush=True)
            if status in ("complete", "failed"):
                return body
            await asyncio.sleep(20)
    raise TimeoutError("Timed out waiting for the score.")


async def score_ad_async(media: str, window_seconds: int = 15, quiet: bool = False) -> dict:
    """Pay, submit, and wait. Returns the final result body (score, label, verdict...)."""
    job = await submit_ad(media, window_seconds)
    if not quiet:
        print(f"Paid & queued: id={job['id']} (scoring takes ~5-8 min)", flush=True)
    return await wait_for_score(job["status_url"], quiet=quiet)


def score_ad(media: str, window_seconds: int = 15, quiet: bool = False) -> dict:
    """Synchronous wrapper — safe to call from LLM tool handlers."""
    return asyncio.run(score_ad_async(media, window_seconds, quiet=quiet))
